Introspect Windows CI image and skip setup accordingly

mokagio opened the pull request:
https://linear.app/a8c/issue/AINFRA-1459

---

- [ ] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

mokagio committed:
Remove some trailing whitespaces

mokagio committed:
Add Windows CI image check and bare bone conditional steps skip

dangermattic commented:
<!--
  0 Errors
  1 Warning: Please add an entry in the <co...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the <code>CHANGELOG.md</code> file to describe the changes made by this PR</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Move Windows developer mode instruction in env setup block

mokagio committed:
Only attempt to setup Windows 10 SDK when not in custom CI

AliSoftware committed:
Extract Windows cert install to dedicated script

AliSoftware commented on tests/test_prepare_windows_host_for_app_distribution.ps1:
```diff
@@ -141,6 +141,67 @@ if ($output -match [regex]::Escape($expectedNoFileSkipMessage)) {
+$ciDir = "C:\CI"
+if (-not (Test-Path $ciDir)) {
+  New-Item -ItemType Directory -Path $ciDir | Out-Null
+}
```
@mokagio Note: now that I've updated the logic to expose the `ami_version` in the `AMI_VERSION` env var instead of in a file in https://github.com/Automattic/buildkite-ci/pull/740/commits/d5495b0b582d5d164f9c05305129bf88bc899ba9, I've [updated the ps1 script accordingly above](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/192/files#diff-1c0800a5a0a1ccef42992bbb8a96402945d8549d7b5c06c4df3ccf6a80d597cbR30-R36) but haven't touched the unit test, so that part will still have to be updated accordingly before merging.

mokagio commented on bin/setup_windows_code_signing.ps1:
```diff
@@ -0,0 +1,22 @@
+# Installs the code signing certificate for Windows app distribution.
+#
+# The certificate is stored in our AWS SecretsManager storage (`windows-code-signing-certificate` secret ID)
```
```suggestion
# The certificate is stored in our AWS SecretsManager storage (`windows-code-signing-certificate` secret ID).
```

mokagio committed:
Minor grammar fix: Add missing period

mokagio committed:
Update Windows unit tests to match new env var check

mokagio committed:
Make `setup_windows_code_signing.ps1` executable

See unit tests failure:
https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/1294/steps/table?sid=019a3812-89c0-4a6b-a3e8-a99250b72fcf#019a3812-8a30-4e50-9adc-362762c918ca/129-162

mokagio committed:
Use `$env:` instead of `[Environment]::Set...`

mokagio commented on bin/prepare_windows_host_for_app_distribution.ps1:
```diff
@@ -27,6 +27,14 @@ param (
+if (![string]::IsNullOrWhiteSpace($amiVersion) -and [version]$amiVersion -ge [version]'0.2') {
+  Write-Output "Tooling is already pre-installed in AMI versions >= 0.2. Only installing code signing certificate."
+  & "setup_windows_code_signing.ps1"
+  Exit 0
```
I wonder if `Exit $LastExitCode` would be more appropriate here. 

I remember looking into the exit code management before sabbatical, but I don't remember what I learned or if/where I tracked my findings.

Logged https://linear.app/a8c/issue/AINFRA-1466/audit-ci-toolkit-ps1-scripts-for-appropriate-exit-code-reporting to followup.

mokagio commented on bin/prepare_windows_host_for_app_distribution.ps1:
```diff
@@ -27,6 +27,14 @@ param (
 # Stop script execution when a non-terminating error occurs
 $ErrorActionPreference = "Stop"
 
+$amiVersion = [Environment]::GetEnvironmentVariable('AMI_VERSION', 'Machine')
+Write-Output "AMI_VERSION is: $amiVersion"
+if (![string]::IsNullOrWhiteSpace($amiVersion) -and [version]$amiVersion -ge [version]'0.2') {
+  Write-Output "Tooling is already pre-installed in AMI versions >= 0.2. Only installing code signing certificate."
+  & "setup_windows_code_signing.ps1"
+  Exit 0
+}
```
Eventually, this script should contain only this call. Tracked https://linear.app/a8c/issue/AINFRA-1465/remove-env-setup-logic-now-in-ami-from-ci-toolkit-once-all-projects for it.

AliSoftware replied:
Actually I think ideally we should even just eventually remove that script and have client projects call `& setup_windows_code_signing.ps1` directly themselves 🙃 (and only if they really need the distribution certificate; eg Unit Tests jobs probably won't even need it?)

mokagio committed:
Run tests in the same process as where $env was set

mokagio committed:
Temporarily disable test for AMI env var check

Having issues setting the env in a transient way but that the scripts
can read.

mokagio committed:
Add changelog entry for conditional Windows behavior

mokagio committed:
Merge branch 'trunk' into ainfra-1459-update-ci-toolkit-to-introspect-ci-env-and-skip-unnecessary

AliSoftware commented on bin/prepare_windows_host_for_app_distribution.ps1:
```diff
@@ -115,29 +115,24 @@ if ($developerMode.State -eq 'Enabled') {
-    Write-Output "[!] Failed to download code signing certificate."
-    Exit $LastExitCode
-  }
+  & "setup_windows_code_signing.ps1"
```
Should we add a test for exit code here? I.e.

```suggestion
  & "setup_windows_code_signing.ps1"
  If ($LastExitCode -ne 0) {
    Exit $LastExitCode
  }
```
Or is that supposed to be automatically handled implicitly in that case?

mokagio replied:
Similar to the earlier question regarding exit codes, I'm not sure. 

Will look at it as part of [linear.app/a8c/issue/AINFRA-1466/audit-ci-toolkit-ps1-scripts-for-appropriate-exit-code-reporting](https://linear.app/a8c/issue/AINFRA-1466/audit-ci-toolkit-ps1-scripts-for-appropriate-exit-code-reporting).
